src/sdk/pynni/nni/compression/tensorflow/compressor.py

```diff
@@ -1,204 +1,300 @@
 
-        The model will be instrumented and user should never edit it after calling this method.
-        """
-        if self.modules_to_compress is None:
-            self.modules_to_compress = []
-            for keras_layer in self.bound_model.layers:
-                layer = LayerInfo(keras_layer)
-                config = self.select_config(layer)
-                if config is not None:
-                    self.modules_to_compress.append((layer, config))
-        return self.modules_to_compress
+        _instrument_model(model, self.wrappers)
 
-    def compress(self):
+    def set_wrappers_attribute(self, name, value):
         """
-        Compress the model with algorithm implemented by subclass.
-
-        The model will be instrumented and user should never edit it after calling this method.
-        `self.modules_to_compress` records all the to-be-compressed layers
+        Call ``setattr`` on all wrappers.
         """
-        modules_to_compress = self.detect_modules_to_compress()
-        for layer, config in modules_to_compress:
-            self._instrument_layer(layer, config)
-        return self.bound_model
+        for wrapper in self.wrappers:
+            setattr(wrapper, name, value)
 
-    def get_modules_to_compress(self):
-        """
-        To obtain all the to-be-compressed layers.
 
-        Returns
-        -------
-        self.modules_to_compress : list
-            a list of the layers, each of which is a tuple (`layer`, `config`),
-            `layer` is `LayerInfo`, `config` is a `dict`
-        """
-        return self.modules_to_compress
+class Pruner(Compressor):
+    """
+    Base class for pruning algorithms.
 
-    def select_config(self, layer):
-        """
-        Find the configuration for `layer` by parsing `self.config_list`
+    End users should use ``compress`` and callback APIs to prune their models.
 
-        Parameters
-        ----------
-        layer: LayerInfo
-            one layer
+    The underlying model is instrumented upon initialization of pruner object.
+    So if you want to pre-train the model, train it before creating pruner object.
 
-        Returns
-        -------
-        ret : config or None
-            the retrieved configuration for this layer, if None, this layer should
-            not be compressed
-        """
-        ret = None
-        if layer.type is None:
-            return None
-        for config in self.config_list:
-            config = config.copy()
-            config['op_types'] = self._expand_config_op_types(config)
-            if layer.type not in config['op_types']:
-                continue
-            if config.get('op_names') and layer.name not in config['op_names']:
-                continue
-            ret = config
-        if ret is None or ret.get('exclude'):
-            return None
-        return ret
+    The compressed model can only execute in eager mode.
```

Comment:
is this statement still true?

Reply:
Yes, we apply masks by modifying weight variables, which can only be done in eager mode.

Reply:
so the performance will be affected if we apply masks on the model? how much performance degradation it would induce?

Reply:
I don't know...

Reply:
better to check the performance degradation using simple test